extensions/control-plane/transfer/transfer-pull-http-dynamic-receiver/src/main/java/org/eclipse/edc/connector/receiver/http/dynamic/HttpDynamicEndpointDataReferenceReceiver.java

```diff
@@ -0,0 +1,163 @@
+import org.eclipse.edc.spi.types.TypeManager;
+import org.eclipse.edc.spi.types.domain.edr.EndpointDataReference;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import static dev.failsafe.Failsafe.with;
+import static java.lang.String.format;
+import static org.eclipse.edc.util.string.StringUtils.isNullOrBlank;
+
+/**
+ * Implementation of a {@link EndpointDataReferenceReceiver} that posts
+ * the {@link EndpointDataReference} to an existing http endpoint sent by clients when initiating the data transfer.
+ */
+public class HttpDynamicEndpointDataReferenceReceiver implements EndpointDataReferenceReceiver {
+
+
+    public static final String HTTP_RECEIVER_ENDPOINT = "receiver.http.endpoint";
+
+    private static final MediaType JSON = MediaType.get("application/json");
+
+    private Monitor monitor;
+    private OkHttpClient httpClient;
+    private TypeManager typeManager;
+    private RetryPolicy<Object> retryPolicy;
+
+    private TransferProcessStore transferProcessStore;
+
+    private String authKey;
+    private String authToken;
+
+    private HttpDynamicEndpointDataReferenceReceiver() {
+    }
+
+    @Override
+    public CompletableFuture<Result<Void>> send(@NotNull EndpointDataReference edr) {
+
+        var processId = transferProcessStore.processIdForDataRequestId(edr.getId());
+
+        if (processId == null) {
+            throw new EdcException(format("Failed to found processId for DataRequestId %s", edr.getId()));
+        }
+        var transferProcess = transferProcessStore.find(processId);
+
+        if (transferProcess == null) {
+            throw new EdcException(format("Failed to found transfer process for id %s", processId));
+        }
+
+        var endpoint = transferProcess.getProperties().get(HTTP_RECEIVER_ENDPOINT);
+
+        if (endpoint != null) {
+            return sendEdr(edr, endpoint);
+        } else {
+            monitor.debug(format("Missing %s property in the transfer process properties", HTTP_RECEIVER_ENDPOINT));
+            return CompletableFuture.completedFuture(Result.success());
+        }
+
+
+    }
+
+    @NotNull
+    private CompletableFuture<Result<Void>> sendEdr(@NotNull EndpointDataReference edr, String endpoint) {
+
+
+        var requestBody = RequestBody.create(typeManager.writeValueAsString(edr), JSON);
+        var requestBuilder = new Request.Builder().url(endpoint).post(requestBody);
+        if (!isNullOrBlank(authKey) && !isNullOrBlank(authToken)) {
+            requestBuilder.header(authKey, authToken);
+        }
+        try (var response = with(retryPolicy).get(() -> httpClient.newCall(requestBuilder.build()).execute())) {
+            if (response.isSuccessful()) {
+                var body = response.body();
+                if (body == null) {
+                    throw new EdcException(format("Received empty response body when receiving endpoint data reference at uri: %s", endpoint));
```

Comment:
Shouldn't this return a failed future? Same as line 105

Reply:
Indeed it should